
This patch permits to gain many time in pick color. There is additionnal informations here :

http://www.vtk.org/Bug/view.php?id=7111

Index: Rendering/vtkOpenGLDisplayListPainter.cxx
===================================================================
RCS file: /cvsroot/VTK/VTK/Rendering/vtkOpenGLDisplayListPainter.cxx,v
retrieving revision 1.6
diff -u -r1.6 vtkOpenGLDisplayListPainter.cxx
--- Rendering/vtkOpenGLDisplayListPainter.cxx	17 Jan 2008 21:01:02 -0000	1.6
+++ Rendering/vtkOpenGLDisplayListPainter.cxx	13 Jun 2008 12:40:16 -0000
@@ -79,15 +79,21 @@
     return;
     }
 
+    long time = this->GetPreviousRenderingForSelection();
+    if(time == 0)
+        time = this->BuildTime;
+    else
+      time = renderer->GetRenderTime();
+
   this->TimeToDraw = 0.0;
 
   // if something has changed regenrate display lists.
   if (!this->DisplayListId || 
-    this->GetMTime() > this->BuildTime ||
-    this->GetInput()->GetMTime() > this->BuildTime ||
-    actor->GetProperty()->GetMTime() > this->BuildTime ||
+    (this->GetMTime() > time && time == this->BuildTime) ||
+    this->GetInput()->GetMTime() > time ||
+    actor->GetProperty()->GetMTime() > time ||
     renderer->GetRenderWindow() != this->LastWindow.GetPointer() ||
-    this->Information->GetMTime() > this->BuildTime || 
+    //this->Information->GetMTime() > this->BuildTime || 
     this->LastUsedTypeFlags != typeflags)
     {
     this->ReleaseList();
@@ -96,11 +102,10 @@
     // generate the display list.
     this->Superclass::RenderInternal(renderer, actor, typeflags);
     glEndList();
-
+    }
     this->BuildTime.Modified();
     this->LastWindow = renderer->GetRenderWindow();
     this->LastUsedTypeFlags = typeflags;
-    }
 
   // Time the actual drawing.
   this->Timer->StartTimer();
Index: Rendering/vtkPainter.cxx
===================================================================
RCS file: /cvsroot/VTK/VTK/Rendering/vtkPainter.cxx,v
retrieving revision 1.6
diff -u -r1.6 vtkPainter.cxx
--- Rendering/vtkPainter.cxx	2 Jun 2008 17:42:38 -0000	1.6
+++ Rendering/vtkPainter.cxx	13 Jun 2008 12:40:16 -0000
@@ -85,6 +85,8 @@
   this->SetInformation(temp);
   temp->Delete();
 
+  this->PreviousRenderingForSelection = 0;
+
   vtkPainter::STATIC_DATA()->Set(this->Information, 0);
   vtkPainter::CONSERVE_MEMORY()->Set(this->Information, 0);
   vtkPainter::HIGH_QUALITY()->Set(this->Information, 1);
@@ -211,6 +213,9 @@
 
   this->PrepareForRendering(renderer, actor);
   this->RenderInternal(renderer, actor, typeflags);
+
+  // If the previous rendering selection information exist remove it
+  SetPreviousRenderingForSelection(0);
 }
 
 //-----------------------------------------------------------------------------
Index: Rendering/vtkPainter.h
===================================================================
RCS file: /cvsroot/VTK/VTK/Rendering/vtkPainter.h,v
retrieving revision 1.6
diff -u -r1.6 vtkPainter.h
--- Rendering/vtkPainter.h	2 Jun 2008 17:42:38 -0000	1.6
+++ Rendering/vtkPainter.h	13 Jun 2008 12:40:16 -0000
@@ -148,6 +148,21 @@
   virtual vtkDataObject* GetOutput()
     { return this->Input; }
 
+  // Description:
+  // If you're making a selection rendering please add the TimeStamp time of the
+  // selection rendering.
+  // If the value is equal to zero then the previous rendering is not a selection rendering.
+  // If the value is different of zero then the previous rendering was a selection rendering.
+  void SetPreviousRenderingForSelection(int value)
+  {
+    this->PreviousRenderingForSelection = value;
+  }
+
+  int GetPreviousRenderingForSelection()
+  {
+    return this->PreviousRenderingForSelection;
+  }
+
 //BTX
 protected:
   vtkPainter();
@@ -235,6 +250,8 @@
 
   double TimeToDraw;
   vtkTimerLog* Timer;
+
+  int PreviousRenderingForSelection;
   
   vtkWeakPointer<vtkWindow> LastWindow; // Window used for previous render.
                          // This is not reference counted.
Index: Rendering/vtkRenderer.cxx
===================================================================
RCS file: /cvsroot/VTK/VTK/Rendering/vtkRenderer.cxx,v
retrieving revision 1.240
diff -u -r1.240 vtkRenderer.cxx
--- Rendering/vtkRenderer.cxx	16 Apr 2008 18:46:05 -0000	1.240
+++ Rendering/vtkRenderer.cxx	13 Jun 2008 12:40:17 -0000
@@ -1863,6 +1863,8 @@
       {
       orig_mapper->SetPainter(orig_painter);
       orig_painter->UnRegister(this);
+      // Inform the painters that this rendering is a selection rendering
+      orig_painter->SetPreviousRenderingForSelection(orig_mapper->GetMTime());
       }
     }
   if (!orig_painter)
Index: Rendering/vtkRenderer.h
===================================================================
RCS file: /cvsroot/VTK/VTK/Rendering/vtkRenderer.h,v
retrieving revision 1.139
diff -u -r1.139 vtkRenderer.h
--- Rendering/vtkRenderer.h	10 Jan 2008 13:51:09 -0000	1.139
+++ Rendering/vtkRenderer.h	13 Jun 2008 12:40:18 -0000
@@ -426,6 +426,8 @@
   // Initial value is false.
   vtkGetMacro(LastRenderingUsedDepthPeeling,int);
   
+  vtkGetMacro(RenderTime,int);
+
 protected:
   vtkRenderer();
   ~vtkRenderer();
